token-minter/sources/modules/mint_stage.move

```diff
@@ -0,0 +1,436 @@
+module minter::mint_stage {
+
+    use std::error;
+    use std::option;
+    use std::option::Option;
+    use std::signer;
+    use std::string::String;
+    use std::vector;
+    use aptos_std::simple_map;
+    use aptos_std::simple_map::SimpleMap;
+    use aptos_std::smart_table::{Self, SmartTable};
+    use aptos_framework::event;
+    use aptos_framework::object;
+    use aptos_framework::object::{ConstructorRef, Object};
+    use aptos_framework::timestamp;
+
+    /// The mint stage start time must be less than the mint stage end time.
+    const EINVALID_START_TIME: u64 = 1;
+    /// The mint stage end time must be greater than the current time.
+    const EINVALID_END_TIME: u64 = 2;
+    /// The mint stage has not started yet.
+    const EMINT_STAGE_NOT_STARTED: u64 = 3;
+    /// The mint stage has ended.
+    const EMINT_STAGE_ENDED: u64 = 4;
+    /// The mint stage allowlist balance is insufficient.
+    const EINSUFFICIENT_ALLOWLIST_BALANCE: u64 = 5;
+    /// The user is not allowlisted.
+    const EUSER_NOT_ALLOWLISTED: u64 = 6;
+    /// The caller is not the owner of the mint stage.
+    const ENOT_OWNER: u64 = 7;
+    /// The mint stage does not exist.
+    const ESTAGE_DOES_NOT_EXIST: u64 = 8;
+    /// The mint stage data does not exist.
+    const EMINT_STAGE_DATA_DOES_NOT_EXIST: u64 = 9;
+    /// The mint stage max per user balance is insufficient.
+    const EINSUFFICIENT_MAX_PER_USER_BALANCE: u64 = 10;
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct MintStageData has key {
+        /// stage ("private", "presale", "public", etc.) to mint stage mapping.
+        mint_stages: SimpleMap<String, MintStage>,
+        /// Stages sorted from the earliest start time.
+        stages: vector<String>,
+    }
+
+    struct MintStage has store {
+        /// The start time of the mint stage.
+        start_time: u64,
+        /// The end time of the mint stage.
+        end_time: u64,
+        /// Allowlist of addresses with their mint allowance.
+        /// Empty allowlist means there are no restrictions.
+        allowlist: SmartTable<address, u64>,
+        /// If allowlist is empty, then everyone can mint but configurable with an optional max limit.
+        no_allowlist: NoAllowlist,
```

Comment:
does this work with that some users are managed by the allowlist, but others are in the no_allowlist? so that `allowlist` and `no_allowlist` coexist and both needed to be checked against when checking for minting.

Reply:
My design is that they should not coexist. If a user wants to do this, they should create a separate stage, i.e. have 2 stages, one for allowlist and one for no allowlist. 

We provide `assert_active_and_execute` function for them to choose to execute if they have more complex logic, rather than calling `execute_earliest_stage`.

Reply:
makes sense, sgtm!